substrate/client/authority-discovery/src/worker.rs

```diff
@@ -306,30 +332,50 @@ where
 	fn addresses_to_publish(&self) -> impl Iterator<Item = Multiaddr> {
 		let peer_id: Multihash = self.network.local_peer_id().into();
 		let publish_non_global_ips = self.publish_non_global_ips;
-		let addresses = self.network.external_addresses().into_iter().filter(move |a| {
-			if publish_non_global_ips {
-				return true
-			}
+		let addresses: Vec<_> = {
+			let addresses = match self.public_addresses_only {
+				Some(ref public_addresses) => public_addresses.clone(),
+				None => self
+					.network
+					.external_addresses()
+					.into_iter()
+					.filter(move |a| {
+						if publish_non_global_ips {
+							return true
+						}
+
+						a.iter().all(|p| match p {
+							// The `ip_network` library is used because its `is_global()` method is
+							// stable, while `is_global()` in the standard library currently isn't.
+							multiaddr::Protocol::Ip4(ip) if !IpNetwork::from(ip).is_global() =>
+								false,
+							multiaddr::Protocol::Ip6(ip) if !IpNetwork::from(ip).is_global() =>
+								false,
+							_ => true,
+						})
+					})
+					.collect(),
+			};
 
-			a.iter().all(|p| match p {
-				// The `ip_network` library is used because its `is_global()` method is stable,
-				// while `is_global()` in the standard library currently isn't.
-				multiaddr::Protocol::Ip4(ip) if !IpNetwork::from(ip).is_global() => false,
-				multiaddr::Protocol::Ip6(ip) if !IpNetwork::from(ip).is_global() => false,
-				_ => true,
-			})
-		});
+			// The address must include the peer id if not already set.
+			addresses
+				.into_iter()
+				.map(move |a| {
+					if a.iter().any(|p| matches!(p, multiaddr::Protocol::P2p(_))) {
+						a
+					} else {
+						a.with(multiaddr::Protocol::P2p(peer_id))
+					}
+				})
+				.collect()
+		};
 
-		debug!(target: LOG_TARGET, "Authority DHT record peer_id='{:?}' addresses='{:?}'", peer_id, addresses.clone().collect::<Vec<_>>());
+		debug!(
+			target: LOG_TARGET,
+			"Authority DHT record peer_id='{peer_id:?}' addresses='{addresses:?}'",
```

Comment:
nit: I think we could `Display` the peerID instead of `Debug`, I've seen some strange logs in grafana (my mistake here sorry)

Reply:
`Vec<Multiaddr>` doesn't implement a `Display` trait, so we are left with `Debug` here, unless we want to implement custom formatting.

What was wrong with the logs in grafana?

Reply:
The logs where displaying the digest bytes of the peerID and not the peerID string format, which made this hard to follow

Reply:
Aaah, got it wrong thinking you are suggesting to use `Display` for `Multiaddr`. Totally fine to `Display` peer id, will fix.

Reply:
Done